src/Razor/benchmarks/Microsoft.AspNetCore.Razor.Microbenchmarks/LanguageServer/RazorDocumentMappingBenchmark.cs

```diff
@@ -0,0 +1,179 @@
+    public int Blocks { get; set; }
+
+    [GlobalSetup]
+    public async Task InitializeRazorCSharpFormattingAsync()
+    {
+        EnsureServicesInitialized();
+
+        var projectRoot = Path.Combine(RepoRoot, "src", "Razor", "test", "testapps", "ComponentApp");
+        var projectFilePath = Path.Combine(projectRoot, "ComponentApp.csproj");
+        _filePath = Path.Combine(projectRoot, "Components", "Pages", $"Generated.razor");
+
+        WriteSampleFile(_filePath, Blocks, out var indexes);
+
+        Indexes = indexes;
+
+        var targetPath = "/Components/Pages/Generated.razor";
+
+        DocumentSnapshot = GetDocumentSnapshot(projectFilePath, _filePath, targetPath);
+
+        var codeDocument = await DocumentSnapshot.GetGeneratedOutputAsync();
+        CSharpDocument = codeDocument.GetCSharpDocument();
+    }
+
+    private static void WriteSampleFile(string filePath, int blocks, out List<int> indexes)
+    {
+        indexes = new List<int>();
+        var data = @"
+@{
+    y = 456;
+}
+
+<div>
+    <span>@DateTime.Now</span>
+    @if (true)
+    {
+        var x = 123;
+        <span>@x</span>
+    }
+</div>
+
+@code {
+    public string Prop$INDEX$ { get; set; }
+
+    public string[] SomeList$INDEX$ { get; set; }
+
+    public class Foo$INDEX$
+    {
+        @* This is a Razor Comment *@
+        void Method() { }
+    }
+}
+
+";
+        using var fileStream = File.CreateText(filePath);
+
+        var index = 0;
+        for (var i = 0; i < blocks; i++)
+        {
+            var fileSegment = data.Replace("$INDEX$", i.ToString());
+            fileStream.WriteLine(fileSegment);
+            index += fileSegment.Length;
+            indexes.Add(index);
+        }
+    }
+
+    [Benchmark]
+    public LinePosition Loop()
+    {
+        LinePosition position = default;
+        foreach (var index in Indexes)
+        {
+            TryMapToHostDocumentPosition(CSharpDocument, index, out position, out _);
+        }
+
+        return position;
+    }
+
+    [Benchmark]
+    public LinePosition BinarySearch()
+    {
+        LinePosition position = default;
+        foreach (var index in Indexes)
+        {
+            DocumentMappingService.TryMapToHostDocumentPosition(CSharpDocument, index, out position, out _);
+        }
+
+        return position;
+    }
+
+    // old code, copied from RazorDocumentMappingService before making changes
+    private bool TryMapToHostDocumentPosition(IRazorGeneratedDocument generatedDocument, int generatedDocumentIndex, out LinePosition hostDocumentPosition, out int hostDocumentIndex)
+    {
+        if (generatedDocument is null)
+        {
+            throw new ArgumentNullException(nameof(generatedDocument));
+        }
+
+        if (generatedDocument.CodeDocument is not { } codeDocument)
+        {
+            throw new InvalidOperationException("Cannot use document mapping service on a generated document that has a null CodeDocument.");
+        }
```

Comment:
Do we need to benchmark the argument checking?

Reply:
Do we need to benchmark "Is a binary search better than a loop?"? 😛